uspace/lib/cpp/include/__bits/thread/async.hpp

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2019 Jaroslav Jindrak
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * - Redistributions of source code must retain the above copyright
+ *   notice, this list of conditions and the following disclaimer.
+ * - Redistributions in binary form must reproduce the above copyright
+ *   notice, this list of conditions and the following disclaimer in the
+ *   documentation and/or other materials provided with the distribution.
+ * - The name of the author may not be used to endorse or promote products
+ *   derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef LIBCPP_BITS_THREAD_ASYNC
+#define LIBCPP_BITS_THREAD_ASYNC
+
+#include <__bits/thread/future.hpp>
+#include <__bits/thread/future_common.hpp>
+#include <__bits/thread/shared_state.hpp>
+#include <__bits/type_traits/result_of.hpp>
+#include <__bits/utility/forward_move.hpp>
+
+namespace std
+{
+
+    enum class launch
+    {
+        async = 1,
+        deferred
+    };
+
+    namespace aux
+    {
+        /**
+         * Note: The reason we keep the actual function
+         *       within the aux namespace is that were the non-policy
+         *       version of the function call the other one in the std
+         *       namespace, we'd get resolution conflicts. This way
+         *       aux::async is properly called even if std::async is
+         *       called either with or without a launch policy.
+         */
+        template<class F, class... Args>
+        future<result_of_t<decay_t<F>(decay_t<Args>...)>>
+        async(launch policy, F&& f, Args&&... args)
+        {
+            using result_t = result_of_t<decay_t<F>(decay_t<Args>...)>;
+
+            bool async = (static_cast<int>(policy) &
+                          static_cast<int>(launch::async)) != 0;
+            bool deferred = (static_cast<int>(policy) &
+                             static_cast<int>(launch::deferred)) != 0;
+
+            /**
+             * Note: The case when async | deferred is set in policy
+             *       is implementation defined, feel free to change.
```

Comment:
What do other standard library implementations do? IMO if there is no reason to do otherwise, we should match their behavior for compatibility.

Reply:
I do not know, but I think that does not matter. The outside behaviour of both the deferred and async policies is identical (and we are supposed to choose one here). However, the async policy carries a risk (resource busy - fibril/thread unavailable) while the deferred policy does not (it simply executes the function when the result is retrieved). This IMHO makes the deferred policy superior for a default (no chance of surprises).

Reply:
I see. That makes sense. Could you add that rationale to the comment?

Reply:
Will do.